eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java

```diff
@@ -1072,11 +1088,60 @@ private ClientResponse makeRemoteCall(Action action) throws Throwable {
      * @throws Throwable
      *             on any error.
      */
-    private ClientResponse makeRemoteCall(Action action, int serviceUrlIndex)
-            throws Throwable {
+    private ClientResponse makeRemoteCall(Action action, int serviceUrlIndex) throws Throwable {
+        String serviceUrl;
+        try {
+            serviceUrl = eurekaServiceUrls.get().get(serviceUrlIndex);
+            return makeRemoteCallWithFollowRedirect(action, serviceUrl);
+        } catch (Throwable t) {
+            if (eurekaServiceUrls.get().size() > ++serviceUrlIndex) {
+                logger.warn("Trying backup: " + eurekaServiceUrls.get().get(serviceUrlIndex));
+                SERVER_RETRY_COUNTER.increment();
+                return makeRemoteCall(action, serviceUrlIndex);
+            } else {
+                ALL_SERVER_FAILURE_COUNT.increment();
+                logger.error("Can't contact any eureka nodes - possibly a security group issue?", t);
+                throw t;
+            }
+        }
+    }
+
+    private ClientResponse makeRemoteCallWithFollowRedirect(Action action, String serviceUrl) throws Throwable {
+        URI targetUrl = new URI(serviceUrl);
+        for(int followRedirectCount = 0; followRedirectCount < 10; followRedirectCount++) {
+            ClientResponse clientResponse = makeRemoteCall(action, targetUrl.toString());
+            if(clientResponse.getStatus() < 300) {
```

Comment:
We need to check for code 302. Status codes are analyzed by isOk method
which accepts <200,300) + 302 always, plus 403/404 for certain actions.
I have changed this condition to != 302 in another commit which I will push
soon.

On Mon, Mar 30, 2015 at 4:28 PM, Nitesh Kant <notifications@github.com>
wrote:

> In eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java
> <https://github.com/Netflix/eureka/pull/486#discussion_r27441830>:
>
> > +                logger.warn("Trying backup: " + eurekaServiceUrls.get().get(serviceUrlIndex));
> > +                SERVER_RETRY_COUNTER.increment();
> > +                return makeRemoteCall(action, serviceUrlIndex);
> > +            } else {
> > +                ALL_SERVER_FAILURE_COUNT.increment();
> > +                logger.error("Can't contact any eureka nodes - possibly a security group issue?", t);
> > +                throw t;
> > +            }
> > +        }
> > +    }
> > +
> > +    private ClientResponse makeRemoteCallWithFollowRedirect(Action action, String serviceUrl) throws Throwable {
> > +        URI targetUrl = new URI(serviceUrl);
> > +        for(int followRedirectCount = 0; followRedirectCount < 10; followRedirectCount++) {
> > +            ClientResponse clientResponse = makeRemoteCall(action, targetUrl.toString());
> > +            if(clientResponse.getStatus() < 300) {
>
> Isn't the below code checking for redirects? Shouldn't this check be
> checking for code b/w 300-400 ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Netflix/eureka/pull/486/files#r27441830>.
>